mesonbuild/dependencies/dub.py

```diff
@@ -59,158 +58,316 @@ def __init__(self, name: str, environment: 'Environment', kwargs: T.Dict[str, T.
         mlog.debug('Determining dependency {!r} with DUB executable '
                    '{!r}'.format(name, self.dubbin.get_path()))
 
+        # if an explicit version spec was stated, use this when querying Dub
+        main_pack_spec = name
+        if 'version' in kwargs:
+            main_pack_spec = f'{name}@{kwargs["version"]}'
+
         # we need to know the target architecture
         arch = self.compiler.arch
 
+        # we need to know the build type as well
+        buildtype = str(environment.coredata.get_option(OptionKey('buildtype')))
+        # MESON types: choices=['plain', 'debug', 'debugoptimized', 'release', 'minsize', 'custom'])),
+        # DUB types: debug (default), plain, release, release-debug, release-nobounds, unittest, profile, profile-gc,
+        # docs, ddox, cov, unittest-cov, syntax and custom
+        if buildtype == 'debugoptimized':
+            buildtype = 'release-debug'
+        elif buildtype == 'minsize':
+            buildtype = 'release'
+
         # Ask dub for the package
-        ret, res = self._call_dubbin(['describe', name, '--arch=' + arch])
+        describe_cmd = [
+            'describe', main_pack_spec, '--arch=' + arch,
+            '--build=' + buildtype, '--compiler=' + self.compiler.get_exelist()[-1]
+        ]
+        ret, res, err = self._call_dubbin(describe_cmd)
 
         if ret != 0:
+            mlog.debug('DUB describe failed: ' + err)
+            if 'locally' in err:
+                mlog.error(mlog.bold(main_pack_spec), 'is not present locally. You may try the following command:')
+                mlog.log(mlog.bold('dub fetch ' + main_pack_spec))
             self.is_found = False
             return
 
-        comp = self.compiler.get_id().replace('llvm', 'ldc').replace('gcc', 'gdc')
-        packages = []
+        # A command that might be useful in case of missing DUB package
+        def dub_build_deep_command() -> str:
+            return f'dub run dub-build-deep --yes -- "{main_pack_spec}" ' + \
+                f'--arch={arch} --compiler={self.compiler.get_exelist()[-1]} ' + \
+                f'--build={buildtype}'
```

Comment:
`dub-build-deep` is a dub package I wrote to build the missing dependencies of a package.
https://github.com/rtbo/dub-build-deep

Reply:
I assume that it's impossible to end up with hilarious errors like main_pack_spec having a double quote char, but compiler.get_exelist can theoretically have whitespace, probably? Depends where it is installed?

You can let python do the quoting for both by using `{expr!r}` assuming they are guaranteed strings. Alternatively, define the command as a list, then use `shlex.join()` as the string.

Reply:
none of `{expr!r}' and `shlex.join()` are portable on Windows because the escape with single quote.
I went for the following (adhoc double quote escape):
```py
        # A command that might be useful in case of missing DUB package
        def dub_build_deep_command() -> str:
            assert '"' not in main_pack_spec, 'package name cannot contain double quote'
            return f'dub run dub-build-deep --yes -- "{main_pack_spec}" ' + \
                f'--arch={arch} "--compiler={self.compiler.get_exelist()[-1]}" ' + \
                f'--build={buildtype}'
```
The double quote should work everywhere as long as there is no double quote in the string (which should not be possible).
This command is only shown to the user, so it is not highly critical if it can't run as is.

Reply:
... right, I forgot about that.

`mesonlib.quote_arg(str)` or `mesonlib.join_args(list)` should work on either Windows or Linux, though. It's used in calculating command lines for build.ninja so it must always be implemented correctly for the current OS or the resulting generated build would be broken.